packages/playwright-core/src/server/injected/ariaSnapshot.ts

```diff
@@ -0,0 +1,240 @@
+  rootElement = rootElement || injectedScript.document.body;
+  const result = toAriaNode(rootElement);
+  const ariaRoot = result?.ariaNode || { role: '' };
+  try {
+    visit(ariaRoot, rootElement);
+  } finally {
+    injectedScript.utils.endAriaCaches();
+  }
+
+  normalizeStringChildren(ariaRoot);
+  return ariaRoot;
+}
+
+export function renderedAriaTree(injectedScript: InjectedScript, rootElement?: Element): string {
+  return renderAriaTree(injectedScript, generateAriaTree(injectedScript, rootElement));
+}
+
+function normalizeStringChildren(rootA11yNode: AriaNode) {
+  const flushChildren = (buffer: string[], normalizedChildren: (AriaNode | string)[]) => {
+    if (!buffer.length)
+      return;
+    const text = normalizeWhitespace(buffer.join(''));
+    if (text.trim())
+      normalizedChildren.push(text.trim());
+    buffer.length = 0;
+  };
+
+  const visit = (ariaNode: AriaNode) => {
+    const normalizedChildren: (AriaNode | string)[] = [];
+    const buffer: string[] = [];
+    for (const child of ariaNode.children || []) {
+      if (typeof child === 'string') {
+        buffer.push(child);
+      } else {
+        flushChildren(buffer, normalizedChildren);
+        visit(child);
+        normalizedChildren.push(child);
+      }
+    }
+    flushChildren(buffer, normalizedChildren);
+    ariaNode.children = normalizedChildren.length ? normalizedChildren : undefined;
+  };
+  visit(rootA11yNode);
+}
+
+const leafRoles = new Set([
+  'alert', 'blockquote', 'button', 'caption', 'checkbox', 'code', 'columnheader',
+  'definition', 'deletion', 'emphasis', 'generic', 'heading', 'img', 'insertion',
+  'link', 'menuitem', 'menuitemcheckbox', 'menuitemradio', 'meter', 'none', 'option',
+  'presentation', 'progressbar', 'radio', 'rowheader', 'scrollbar', 'searchbox', 'separator',
+  'slider', 'spinbutton', 'strong', 'subscript', 'superscript', 'switch', 'tab', 'term',
+  'textbox', 'time', 'tooltip'
+]);
+
+const normalizeWhitespace = (text: string) => text.replace(/[\s\n]+/g, ' ');
+
+function escapeRegex(string: string) {
+  return string.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
+}
+
+export function matchesAriaTree(injectedScript: InjectedScript, rootElement: Element, template: AriaTemplateNode): { matches: boolean, received: string } {
+  const root = generateAriaTree(injectedScript, rootElement);
+  const matches = nodeMatches(root, template);
+  return { matches, received: renderAriaTree(injectedScript, root) };
+}
+
+function matchesNode(node: AriaNode | string, template: AriaTemplateNode | AriaTemplateString, depth: number): boolean {
+  if (typeof node === 'string' && template.kind === 'string') {
+    const pattern = template.chunks.map(s => typeof s === 'string' ? escapeRegex(normalizeWhitespace(s)) : s.source);
+    return !!node.match(new RegExp(pattern.join('')));
+  }
+
+  if (typeof node === 'object' && template.kind === 'node') {
+    if (template.role && template.role !== node.role)
```

Comment:
How could there be no `role` in the template?

Reply:
That's a regex template